src/replayq_mem_ets_shared.erl

```diff
@@ -0,0 +1,168 @@
+        {write_concurrency, true}
+    ]),
+    ok.
+
+%% @doc Create a new queue.
+-spec new(_) -> queue().
+new(Opts) ->
+    Owner = maps:get(owner, Opts, self()),
+    ok = replayq_registry:register_slot_owner(Owner),
+    #{next_id => 1, owner => Owner}.
+
+%% @doc Peek all items from the queue.
+-spec peek_all(queue()) -> [item()].
+peek_all(Q) ->
+    Owner = owner(Q),
+    Ms = ets:fun2ms(fun({?KEY(Pid, '_'), Item}) when Pid =:= Owner -> Item end),
+    ets:select(?ETS_TAB, Ms).
+
+%% @doc Peek the front item from the queue.
+-spec peek(queue()) -> empty | {value, item()}.
+peek(Q) ->
+    Owner = owner(Q),
+    case first_key(Owner) of
+        {key, Key} ->
+            Item = ets:lookup_element(?ETS_TAB, Key, 2),
+            {value, Item};
+        empty ->
+            empty
+    end.
+
+%% @doc Check if a queue is empty.
+-spec is_empty(queue()) -> boolean().
+is_empty(Q) ->
+    Owner = owner(Q),
+    empty =:= first_key(Owner).
+
+%% @doc Enqueue an item into the queue.
+-spec in(item(), queue()) -> queue().
+in(Item, #{next_id := NextId} = Q) ->
+    Owner = owner(Q),
+    Key = ?KEY(Owner, NextId),
+    ets:insert(?ETS_TAB, {Key, Item}),
+    Q#{next_id => NextId + 1}.
+
+%% @doc Enqueue an item into the queue in reverse order.
+-spec in_r(item(), queue()) -> queue().
+in_r(Item, Q) ->
+    Owner = owner(Q),
+    case first_key(Owner) of
+        {key, ?KEY(Owner, Id)} ->
+            PrevKey = ?KEY(Owner, Id - 1),
+            ets:insert(?ETS_TAB, {PrevKey, Item}),
+            Q;
+        empty ->
+            in(Item, Q)
+    end.
+
+%% @doc Enqueue a batch of items into the queue.
+-spec in_batch([item()], queue()) -> queue().
+in_batch(Items, #{next_id := NextId} = Q) ->
+    Owner = owner(Q),
+    {NewNextId, Inserts} = lists:foldl(
+        fun(Item, {Id, Acc}) -> {Id + 1, [{?KEY(Owner, Id), Item} | Acc]} end,
+        {NextId, []},
+        Items
+    ),
+    ets:insert(?ETS_TAB, Inserts),
+    Q#{next_id => NewNextId}.
+
+%% @doc Dequeue an item from the queue.
+-spec out(queue()) -> {empty, queue()} | {{value, item()}, queue()}.
+out(Q) ->
+    Owner = owner(Q),
+    case first_key(Owner) of
+        {key, ?KEY(Owner, Id)} ->
+            Item = ets:lookup_element(?ETS_TAB, ?KEY(Owner, Id), 2),
+            ets:delete(?ETS_TAB, ?KEY(Owner, Id)),
+            {{value, Item}, Q};
+        empty ->
+            {empty, Q}
+    end.
+
+%% @doc Destroy the queue.
+-spec destroy(queue()) -> ok.
+destroy(Q) ->
+    Owner = owner(Q),
+    ok = purge_by_owner(Owner),
+    ok = replayq_registry:deregister_slot_owner(Owner).
+
+%% @doc Purge the queue by owner pid.
+-spec purge_by_owner(pid()) -> ok.
+purge_by_owner(Owner) ->
+    Ms = ets:fun2ms(fun({?KEY(Pid, '_'), _}) when Pid =:= Owner -> true end),
+    ets:select_delete(?ETS_TAB, Ms),
+    ok.
+
+%% Get the first key in the queue.
+-spec first_key(pid()) -> empty | {key, key()}.
+first_key(Owner) ->
+    case ets:next(?ETS_TAB, ?KEY(Owner, 0)) of
```

Comment:
I think we can't rely on `0` being less than any existing key, given that `in_r` always decreases 1 from the "first" key.